core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java

```diff
@@ -7204,6 +7227,146 @@ protected SqlNode expandDynamicStar(SqlIdentifier id, SqlIdentifier fqId) {
+    // a subset of the select items for an alias. IE:
+    //    SELECT A, B, C from ...
+    // If we're trying to find aliases for "B" we should only check every select statement that
+    // precedes B. IE, the following statement should fail since the "x" alias
+    // does not precede the select item.:
+    //    SELECT A, x, C as x from ...
+    // but the following statement will work:
+    //    SELECT C as x, x, A from ...
+    //
+    // Note that we must also handle the following case (which is allowed in SF):
+    //    SELECT A as x, x as x from ...
+
+    // This argument instructs the expander to only check elements in the select list
+    // up to but not including this index.
+    Integer maxIdx;
+
+    ExtendedAliasExpander(SqlValidatorImpl validator, SqlValidatorScope scope,
+        SqlSelect select) {
+      // This an alias in the WHERE/ON clause, since we need to check all elements
+      // of the select list for aliases
+      super(validator, scope);
+      this.aliasSourceSelect = select;
+      this.maxIdx = select.getSelectList().size();
+    }
+
+
+    ExtendedAliasExpander(SqlValidatorImpl validator, SqlValidatorScope scope,
+        SqlSelect select, Integer maxIdx) {
+      // This constructor is used when attempting to expand a select item, since we need to check
+      // a subset of the select list for alias information
+      // Note that it is also called recursivley, in order to handle expanding chains of aliases,
+      // IE; A as x, x as y, y as z ... etc.
+      super(validator, scope);
+      this.aliasSourceSelect = select;
+      this.maxIdx = maxIdx;
+      assert maxIdx <= select.getSelectList().size() && maxIdx >= 0;
+    }
+
+
+
+    @Override public @Nullable SqlNode visit(SqlIdentifier id) {
+      if (id.isSimple()) {
+        try {
+          // If we can resolve this in the regular scope, we should
+          // since we prioritize all id's that could reference columns in the "from/on" clauses
+          // before we look for aliases within the select list itself.
+          SqlNode sqlNode = super.visit(id);
+          return sqlNode;
+        } catch (Exception e) {
+
+          final SqlNameMatcher nameMatcher =
+              validator.catalogReader.nameMatcher();
+          String name = id.getSimple();
+
+          Pair<Integer, Pair<@Nullable SqlNode, Integer>> occurenceNumExprAndExprIdx =
+              findAliasesInSelect(
+              name, nameMatcher,
+              this.aliasSourceSelect, this.maxIdx);
+
+          int numOccurrencesOfAlias = occurenceNumExprAndExprIdx.left;
+
+
+          if (numOccurrencesOfAlias == 0) {
+            //If we can't find any aliases in the select list, just throw
+            // the original error.
+            throw e;
+          } else if (numOccurrencesOfAlias > 1) {
+            // More than one column has this alias.
+            throw validator.newValidationError(id,
```

Comment:
Adding a comment to make sure this is tested.

Reply:
tested in SqlValidatorTest (see testFailsAmbiguous and testFailsAmbiguous2)